### ISSUE TITLE
Tool to create a random file name

### ISSUE THREAD

**Issue by kriegaex** [spockframework/spock#1448]
### Is your feature request related to a problem?

PR #1199 introduced the `@TempDir` field annotation and corresponding `TempDirExtension`, closing requirement #1126.

Sometimes, having a unique temporary directory is not enough, but unique file names within that directory are also required. For example, when using a `@Shared @TempDir` field in `where:` blocks of multiple features, we still might get name clashes in parallel execution mode, if each feature uses fixed file names. A non-shared or non-static field is not an option when using `where:`, so we have a potential concurrency problem with regard to the shared temporary folder here.

Therefore, it would be helpful if Spock could create randomised temporary file names out of the box. I really mean **just names, not the files themselves.** For the latter, we could use `java.nio.file.Files#createTempFile(java.nio.file.Path, java.lang.String, java.lang.String, java.nio.file.attribute.FileAttribute<?>...)`, but firstly that looks clunky in an example like this:

```groovy
  @TempDir
  @Shared
  Path sharedTempDir

  @Unroll("constructor #iterationIndex")
  def "element size must be positive"() {
    when: "trying to create a random access file with illegal element size"
    createRandomAccessFile()

    then: "an IllegalElementSizeException is thrown"
    thrown FixedElementSizeRandomAccessFile.IllegalElementSizeException

    where:
    createRandomAccessFile << [
      { new ByteArrayRandomAccessFile(Files.createTempFile(sharedTempDir, null, null).toFile().absolutePath, "rw", 0) },
      { new ByteArrayRandomAccessFile(Files.createTempFile(sharedTempDir, null, null).toFile().absolutePath, "rw", -1, 1024) },
      { new ByteArrayRandomAccessFile(Files.createTempFile(sharedTempDir, null, null).toFile(), "rw", -2) },
      { new ByteArrayRandomAccessFile(Files.createTempFile(sharedTempDir, null, null).toFile(), "rw", -99, 1024) }
    ]
  }
```

Secondly, the constructors here are expected to create the files by themselves, not to have them created for them. In this example, they are expected to fail due to illegal parameters, so no file I/O is expected to happen. But just in case the test fails (i.e. the files get created in the constructor anyway, they should be in a temp-dir. Either way, for the constructors I only need random names als `String` or `File` instances, not files.

### Simple case

I was thinking about an annotation-driven extension which can inject temporary files (or optionally just their names), similar to what `@TempDir` does now. Ideally, it could also use a `@TempDir` field for determining the base directory, but that would probably not be so straightforward to implement. The simplest solution could look like that from a user's perspective:

```groovy
@TempFile
def instanceTempFileCreatedForMe
@Shared @TempFile
def sharedTempFileCreatedForMe

@TempFile(create = false)
def instanceTempFileName
@Shared @TempFile(create = false)
def sharedTempFileName
```

Like `@TempDir`, field types `File` and `Path` (default for `def`) would be supported for created temp-files. For `create = false`, only the name would be injected, again either as `File` or `Path`, but here Spock also support `String` and make it the default for `def`.

### Fancy case

If we want to have a fancier solution, something akin to this would be cool, too:

```groovy
@TempDir
def instanceTempDir
@TempDir
def sharedTempDir

@TempFile(dir = "instanceTempDir")
def instanceTempFileCreatedForMe
@Shared @TempFile(dir = "sharedTempDir")
def sharedTempFileCreatedForMe

@TempFile(create = false, dir = "instanceTempDir")
def instanceTempFileName
@Shared @TempFile(create = false, dir = "sharedTempDir")
def sharedTempFileName
```

Besides, an option `create = false` for `@TempDir` would also be nice to have.

In order to the simple solution to work, we could either use a separate extension or extend the existing `TempDirExtension`, supporting a new annotation.

For the latter, fancier case where `@TempDir` refers to a field name annotated by `@TempDir`, probably doing it all in one extension would be the easier way, which is why I am suggesting to extend the existing extension in any case. It is just technically challenging, because of course we cannot simply say `TempDirExtension implements IAnnotationDrivenExtension<TempDir>, IAnnotationDrivenExtension<TempFile>`. One class cannot implement the same interface twice for different generic types. In that case, maybe we would have to refactor the extension to be a global extension, but behaving like an annotation-driven one, similar to `UnrollExtension` handling both `@Unroll` and `@Rollup` as a global extension.

### Direct access to name generator, e.g. from `where:` blocks

Last, but not least, in order to cover the use case in my example code, even something like `@Shared @TempFile(create = false, dir = "sharedTempDir")` would be inadequate, because in the `where:` block we really need multiple file names. This could be covered by the extension exposing a static method creating random file names as strings, paths or files, but without creating physical files, e.g. flavours
  * `File TempFileExtension.tempFile(baseDir)`,
  * `PathTempFileExtension.tempPath(baseDir)`,
  * `StringTempFileExtension.tempName(baseDir)`,
with each oth the flavours also accepting `File`, `Path`, `String` types for base directories. That would involve some `instanceof` checks when implemented in a way compatible to Java 8, but it would offer maximum user comfort.

Then something like this would be possible:

```groovy
org.spockframework.runtime.extension.builtin.TempDirExtension.tempFile
org.spockframework.runtime.extension.builtin.TempDirExtension.tempName

// ...
    where:
    createRandomAccessFile << [
      // test String file name constructors
      { new ByteArrayRandomAccessFile(tempName(sharedTempDir), "rw", 0) },
      { new ByteArrayRandomAccessFile(tempName(sharedTempDir), "rw", -1, 1024) },
      // test File file name constructors
      { new ByteArrayRandomAccessFile(tempFile(sharedTempDir), "rw", -2) },
      { new ByteArrayRandomAccessFile(tempFile(sharedTempDir), "rw", -99, 1024) }
    ]
```

### Describe alternatives you've considered

I looked around the Java, Groovy and Spock APIs, but did not find any obvious solutions. All solutions create the files for me, which is not what I want. I just need the names. If we look e.g. at the code in `java.nio.file.TempFileHelper#generatePath`, we see that the solution is pretty simple, basically `Long.toUnsignedString(n)` where `n` is a random number, plus prefix and suffix, if any:

```java
    // file name generation, same as java.io.File for now
    private static final SecureRandom random = new SecureRandom();
    private static Path generatePath(String prefix, String suffix, Path dir) {
        long n = random.nextLong();
        String s = prefix + Long.toUnsignedString(n) + suffix;
        Path name = dir.getFileSystem().getPath(s);
        // the generated name should be a simple file name
        if (name.getParent() != null)
            throw new IllegalArgumentException("Invalid prefix or suffix");
        return dir.resolve(name);
    }
```

But you see, the methods generating the names in all APIs are private or protected at best, i.e. I cannot use them.

Of course, I can also just implement my own test utility or use a third-party tool library, but unique temp-file names are such a common requirement, I think it makes sense to have it in Spock.

### Additional context

x

**Comment by kriegaex** [comment]
@DQinYuan, you created `TempDirExtension.` Before I start thinking about implementing anything by myself, let me ask you if you are interested in contributing something again.

@leonard84, your thoughts about this requirement would be valuable too, because if you rather feel like rejecting the idea, there is no use starting to hack around.

**Comment by kriegaex** [comment]
FYI, this Groovy helper class is what I am using now. I created it in a "quick & dirty" was, but it is not so bad, I guess. At least it serves my purposes and is versatile:

```groovy
package de.scrum_master.testing

import java.nio.file.Path
import java.security.SecureRandom

class TempFileUtility {
  private static final SecureRandom random = new SecureRandom()

  static Path tempPath(baseDir = "") {
    generatePath("", "", toPath(baseDir))
  }

  static File tempFile(baseDir = "") {
    tempPath(baseDir).toFile()
  }

  static String tempName(baseDir = "") {
    tempFile(baseDir).absolutePath
  }

  private static Path toPath(def pathRepresentation) {
    switch (pathRepresentation) {
      case Path: return pathRepresentation
      case File: return pathRepresentation.toPath()
      default: return new File(pathRepresentation as String).toPath()
    }
  }

  /**
   * Shamelessly copied from private static JRE method
   * {@link java.nio.file.TempFileHelper#generatePath(java.lang.String, java.lang.String, java.nio.file.Path)
   * TempFileHelper.generatePath}
   */
  private static Path generatePath(String prefix, String suffix, Path dir) {
    long n = random.nextLong()
    String s = prefix + Long.toUnsignedString(n) + suffix
    Path name = dir.getFileSystem().getPath(s)
    // the generated name should be a simple file name
    if (name.getParent() != null)
      throw new IllegalArgumentException("Invalid prefix or suffix")
    return dir.resolve(name)
  }
}
```

Usage:
```groovy
import static de.scrum_master.testing.TempFileUtility.tempFile
import static de.scrum_master.testing.TempFileUtility.tempName
// ...

class ByteArrayRandomAccessFileTest extends Specification {
  @TempDir @Shared Path sharedTempDir

  def "element size must be positive"() {
    // ...
    where:
    createRandomAccessFile << [
      { new ByteArrayRandomAccessFile(tempName(sharedTempDir), "rw", 0) },
      { new ByteArrayRandomAccessFile(tempName(sharedTempDir), "rw", -1, 1024) },
      { new ByteArrayRandomAccessFile(tempFile(sharedTempDir), "rw", -2) },
      { new ByteArrayRandomAccessFile(tempFile(sharedTempDir), "rw", -99, 1024) }
    ]
  }
```

No matter whether I need a `String`, `Path` or `File`, I can easily get one, calling the right utility method. And no matter which of the 3 types (or whatever else might represent a path when coerced to `String`) I use as input parameters, it also works.

BTW, in Groovy 4 the tool class could even use the new switch expressions instead of a classical switch-case construct. But this is supposed to work in older Groovy versions, too.

**Comment by leonard84** [comment]
Let me first ask this:

1. how often did you need it, was this the first time or have you had need for this repeatedly in different projects?
2. Why do you want to pass in an Object/Closure to the feature? It makes for really bad reporting. I would suggest to pass in the individual parameters.

```
  @TempDir File sharedTempDir

  def "element size must be positive"() {
    def fileParam = new File(sharedTempDir, parameters[0])
    parameters[0] = (parameters[0] instanceof String) ?  fileParam.absolutePath : fileParam

    when:
	new ByteArrayRandomAccessFile(*parameters)
    // ...
    where:
    parameters << [
 		["aFile.txt", "rw", 0],
 		["aFile.txt", "rw", 0, 1024],
 		[new File("aFile.txt"), "rw", 0],
 		[new File("aFile.txt"), "rw", 0, 1024]
    ]
  }
```

**Comment by kriegaex** [comment]
> 1. how often did you need it, was this the first time or have you had need for this repeatedly in different projects?

Repeatedly. More generally, I need temporary files in many integration tests, but also in some types of tests where it is difficult to isolate the unit under test from filesystem resources due to `java.io` use. In contrast to `java.nio` which could, AFAIK, be mocked away using JimFS, developers do need a native file system in this case. My experience is that, unfortunately, in the wild I mostly meet `java.io`. Like in my examples above, random access files are based on `java.io`. When such classes are directly under test or otherwise an important part of the test setup, mocking them away is counter-productive. Sometimes spying and stubbing on single methods in order to test some corner cases is OK, but in general it makes sense to use file systems during those kinds of tests. The usual workarounds for avoiding problems are to either manually assign or generate unique names or to simply not execute those tests concurrently. The latter is what I see most often, or sometimes developers are just lucky that there are not problems due to their test setups (e.g. sequential execution) and they never thought about it, but the problems are there and usually surface quickly, as soon as tests are being run concurrently. Then everyone starts hacking around, trying to avoid problems with identical file names. so I believe, this problem is worth solving in a more general way.

> 2. Why do you want to pass in an Object/Closure to the feature? It makes for really bad reporting. I would suggest to pass in the individual parameters.

That is great advice. I am still by no means a Groovy pro. In this case, I simply did not remember that it is possible to spread method or constructor arguments. I had seen it before, but not used it much. so the example here is perfectly solvable with `*parameters` and clearly superior to my approach. That would lift the pressure from the extension to offer a directly callable API, even though I think that would still be nice to have. But the main use cases of injectable names still remains as something useful, even though not strictly necessary if using spreaded argument lists. Thanks to Groovy Goodness, this solves most problems. I wonder what the Java people do in that case, other than using reflection or fiddling with method references.

**Comment by kriegaex** [comment]
Just if anyone is reading this, too: We can make spread operator usage a bit smarter. I thought about what @leonard84 said about bad reporting for closures, and it is absolutely true. So let us do it like this:

```groovy
@Unroll("element size #elementSize, buffer size #bufferSize")
def "element size must be positive"() {
  when: "trying to create a random access file with illegal element size"
  new ByteArrayRandomAccessFile(*arguments)

  then: "an IllegalElementSizeException is thrown"
  thrown FixedElementSizeRandomAccessFile.IllegalElementSizeException

  where:
  file              | mode | elementSize | bufferSize
  tempName(tempDir) | "rw" | 0           | "default"
  tempName(tempDir) | "rw" | -1          | 1024
  tempFile(tempDir) | "rw" | -2          | "default"
  tempFile(tempDir) | "rw" | -99         | 1024

  arguments = [file, mode, elementSize, bufferSize] - "default"
}
```

In my IDE, this looks really neat with unrolled method names now:

![image](https://user-images.githubusercontent.com/1537384/158827036-9cdccf06-b22f-4539-9370-ea156c100466.png)

**Comment by leonard84** [comment]
At the current moment I'm not inclined to add something like this to `spock-core`:

1. `@Shared @TempDir` should only be used if the tests are actually sharing the directory, e.g., `@Stepwise`.
2. Creating files in the `where:` block should be avoided and instead moved to the `given:` block, and only passing the necessary parameters
   1. You could pass in the file name
   2. If necessary pass in a closure accepting a file as parameter `{ f -> new File(f, "foo") }` and then call the closure in the `given:` block `fileClosure(tempDir)`.
3. There are many use cases where a random file name doesn't solve the issue as the code you are testing expects certain file names in a directory, e.g., `build.gradle`

**Comment by kriegaex** [comment]
1. Of course. Using a shared directory with random files works just fine in my case.
2. Thanks for the hint again. I know it works. I just hate the ceremony (boilerplate code), because it makes my features less readable and in small features it's more code than the actual feature logic.
3. Yes, but I wanted it in cases in which it does help to solve the issue. You can reject any feature request in the world, saying that it does not solve all problems in the world.

Anyway, I got the message. You don't want it, so I won't create a PR. Time saved.
